File: DeepSpace-Ruby/lib/SpaceCity.rb
# encoding: utf-8

require_relative "SpaceCityToUI"

module Deepspace
class SpaceCity < SpaceStation
    attr_reader :base, :collaborators

    def initialize (base, rest) # base: SpaceStation, rest: Array<SpaceStation>
        createStation(base)
        @collaborators = rest
        @base          = base
    end

    def fire
        disparo = super
        @collaborators.each { |station|
            disparo += station.fire
        }
        disparo
    end

    def protection
        proteccion = super
        @collaboratos.each { |station|
            proteccion += station.protection
        }
        proteccion
    end

    def setLoot (loot)
        super
        Transformation::NOTRANSFORM
    end

    def getUIversion
        SpaceCityToUI.new(self)
    end

    def to_s
        getUIversion.to_s
    end
end
end

File: DeepSpace-Java/src/deepspace/SpaceCity.java
package deepspace;

import java.util.ArrayList;

public class SpaceCity extends SpaceStation {
    private SpaceStation base;
    private ArrayList<SpaceStation> collaborators;

    public SpaceCity (SpaceStation basico, ArrayList<SpaceStation> resto) {
        super(basico);
        this.base = basico;

        for (SpaceStation estacion : resto)
            collaborators.add(estacion);
    }

    public ArrayList<SpaceStation> getCollaborators () {
        return collaborators;
    }

    @Override
    public float fire () {
        float disparo = base.fire();

        for (SpaceStation colaboradores : collaborators)
            disparo += colaboradores.fire();

        return disparo;
    }

    @Override
    public float protection () {
        float muralla = base.protection();

        for (SpaceStation colaboradores : collaborators)
            muralla += colaboradores.protection();

        return muralla;
    }

    public Trasnformation setLoot (Loot loot) {
        base.setLoot(loot);
        return Transformation.NOTRANSFORM;
    }

    @Override
    public String toString() {
        return SpaceCityToUI(this);
    }
}